backend/checkmach.ml

```diff
@@ -27,454 +27,809 @@
-          Unit_info.in_current_unit unit_info fun_name f.fun_dbg;
-          if List.mem (Cmm.Assume S.annotation) f.fun_codegen_options
-          then (
-            report t ~msg:"assumed" ~desc:"fundecl" f.fun_dbg;
-            Unit_info.add_value t.ppf unit_info fun_name Pass)
-          else (
-            (try
-               let _ = check_instr_exn t f.fun_body false in
-               if (not t.unresolved_dependencies)
-                  && not (Unit_info.is_fail unit_info t.fun_name)
-               then (
-                 report t ~msg:"passed" ~desc:"fundecl" f.fun_dbg;
-                 Unit_info.add_value t.ppf unit_info fun_name Pass)
-             with Bail -> debug t Fail);
-            if List.mem (Cmm.Assert S.annotation) f.fun_codegen_options
-            then Unit_info.annotated unit_info fun_name))
+    then Profile.record_call ~accumulate:true ("check " ^ analysis_name) check
 end
 
+(** Check that functions do not allocate on the heap (local allocations are ignored) *)
 module Spec_alloc : Spec = struct
-  let name = "noalloc"
+  let property = Cmm.Noalloc
 
   let enabled () = !Flambda_backend_flags.alloc_check
 
-  let check_callee s (checks : Cmx_format.checks) =
-    String.Set.mem s checks.ui_noalloc_functions
-
-  let add_callee s (checks : Cmx_format.checks) =
-    if not (check_callee s checks)
-    then
-      checks.ui_noalloc_functions
-        <- String.Set.add s checks.ui_noalloc_functions
-
-  let check_specific s = not (Arch.operation_allocates s)
-
-  let annotation = Cmm.Noalloc
+  (* Compact the mapping from function name to Value.t to reduce size of Checks
+     in cmx and memory consumption Compilenv. Different components have
+     different frequencies of Top/Bot. The most likely value is encoded as None
+     (i.e., not stored). *)
+  let encode_return (v : V.t) =
+    match v with Top -> None | Safe -> Some true | Bot -> Some false
+
+  let decode_return = function
+    | None -> V.Top
+    | Some true -> V.Safe
+    | Some false -> V.Bot
+
+  let encode_diverge (v : V.t) =
+    match v with Top -> Some false | Safe -> Some true | Bot -> None
+
+  let decode_diverge = function
+    | None -> V.Bot
+    | Some true -> V.Safe
+    | Some false -> V.Top
+
+  let set_value s (v : Value.t) =
+    let checks = (Compilenv.current_unit_infos ()).ui_checks in
+    let new_value : Checks.value =
+      { nor = encode_return v.nor;
+        exn = encode_return v.exn;
+        div = encode_diverge v.div
+      }
+    in
+    Checks.set_value checks s new_value
+
+  let get_value_opt s =
+    let checks = Compilenv.cached_checks in
+    match Checks.get_value checks s with
+    | None -> None
+    | Some ({ nor; exn; div } : Checks.value) ->
+      Some
+        { Value.nor = decode_return nor;
+          exn = decode_return exn;
+          div = decode_diverge div
+        }
+
+  let transform_specific s ~next ~exn:_ =
+    if Arch.operation_allocates s then Value.transform next else next
```

Comment:
I am surprised this does not also use
`Arch.operation_can_raise`. Arguably,
that will not make a difference at the
moment on amd64...

Reply:
See proposed fix in [commit 5ad2314](https://github.com/ocaml-flambda/flambda-backend/pull/1157/commits/5ad23142261500b52ae605c197854ebad6efe8a7). I'm assuming that specific operations cannot diverge.